### PR TITLE
Retries on HTTP connection drops during reads from S3

### LINKED ISSUE #21957: Read error from S3
**Describe the bug**
When reading from table using diskS3, such exception occurs:

```
Code: 23. DB::Exception: Received from localhost:9000. DB::Exception: Cannot read from istream at offset 0: (while reading column LO_ORDERDATE): (while reading from part ... from mark 14287 with max_rows_to_read = 16384): While executing MergeTreeThread.
```
**Does it reproduce on recent release?**
Yes.

**Stack trace**

```
2021.03.22 14:39:43.010425 [ 65795 ] {} <Error> DB::IBackgroundJobExecutor::jobExecutingTask()::<lambda()>: Code: 23, e.displayText() = DB::Exception: Cannot read from istream at offset 0: (while reading column LO_SUPPKEY): While executing MergeTreeSequentialSource, Stack trace (when copying this message, always include the lines below):

0. Poco::Exception::Exception(std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > const&, int) @ 0xd1610e0 in /essd/CK/s3/programs/clickhouse
1. DB::Exception::Exception(std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > const&, int, bool) @ 0x3ecfda1 in /essd/CK/s3/programs/clickhouse
2. DB::ReadBufferFromIStream::nextImpl() (.cold) @ 0x3c7f652 in /essd/CK/s3/programs/clickhouse
3. DB::ReadBufferFromS3::nextImpl() @ 0xca457c7 in /essd/CK/s3/programs/clickhouse
4. DB::ReadIndirectBufferFromS3::nextImpl() @ 0xae6e517 in /essd/CK/s3/programs/clickhouse
5. DB::SeekAvoidingReadBuffer::nextImpl() @ 0xca509c2 in /essd/CK/s3/programs/clickhouse
6. DB::ReadBuffer::readStrict(char*, unsigned long) @ 0x3f02af1 in /essd/CK/s3/programs/clickhouse
7. DB::CompressedReadBufferBase::readCompressedData(unsigned long&, unsigned long&, bool) @ 0x9c85949 in /essd/CK/s3/programs/clickhouse
8. DB::CompressedReadBufferFromFile::readBig(char*, unsigned long) @ 0x9c86b67 in /essd/CK/s3/programs/clickhouse
9. DB::DataTypeNumberBase<unsigned int>::deserializeBinaryBulk(DB::IColumn&, DB::ReadBuffer&, unsigned long, double) const @ 0x9d66eb3 in /essd/CK/s3/programs/clickhouse
10. DB::IDataType::deserializeBinaryBulkWithMultipleStreams(COW<DB::IColumn>::immutable_ptr<DB::IColumn>&, unsigned long, DB::IDataType::DeserializeBinaryBulkSettings&, std::__1::shared_ptr<DB::IDataType::DeserializeBinaryBulkState>&, std::__1::unordered_map<std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> >, COW<DB::IColumn>::immutable_ptr<DB::IColumn>, std::__1::hash<std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > >, std::__1::equal_to<std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > >, std::__1::allocator<std::__1::pair<std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > const, COW<DB::IColumn>::immutable_ptr<DB::IColumn> > > >*) const @ 0x9d97f38 in /essd/CK/s3/programs/clickhouse
11. DB::MergeTreeReaderCompact::readData(DB::NameAndTypePair const&, COW<DB::IColumn>::immutable_ptr<DB::IColumn>&, unsigned long, unsigned long, unsigned long, bool) @ 0xa4505b2 in /essd/CK/s3/programs/clickhouse
12. DB::MergeTreeReaderCompact::readRows(unsigned long, bool, unsigned long, std::__1::vector<COW<DB::IColumn>::immutable_ptr<DB::IColumn>, std::__1::allocator<COW<DB::IColumn>::immutable_ptr<DB::IColumn> > >&) @ 0xa450a54 in /essd/CK/s3/programs/clickhouse
13. DB::MergeTreeSequentialSource::generate() @ 0xa461ba9 in /essd/CK/s3/programs/clickhouse
14. DB::ISource::work() @ 0xa5af04b in /essd/CK/s3/programs/clickhouse
15. DB::SourceWithProgress::work() @ 0xa70aed7 in /essd/CK/s3/programs/clickhouse
16. void std::__1::__function::__policy_invoker<void ()>::__call_impl<std::__1::__function::__default_alloc_func<DB::PipelineExecutor::addJob(DB::ExecutingGraph::Node*)::'lambda'(), void ()> >(std::__1::__function::__policy_storage const*) @ 0xa5d5bed in /essd/CK/s3/programs/clickhouse
17. DB::PipelineExecutor::executeStepImpl(unsigned long, unsigned long, std::__1::atomic<bool>*) (.constprop.0) @ 0xa5dac24 in /essd/CK/s3/programs/clickhouse
18. DB::PipelineExecutor::executeStep(std::__1::atomic<bool>*) @ 0xa5dc5a9 in /essd/CK/s3/programs/clickhouse
19. DB::PullingPipelineExecutor::pull(DB::Chunk&) @ 0xa5e4475 in /essd/CK/s3/programs/clickhouse
20. DB::PullingPipelineExecutor::pull(DB::Block&) @ 0xa5e4be7 in /essd/CK/s3/programs/clickhouse
21. DB::PipelineExecutingBlockInputStream::readImpl() @ 0xa5d4a78 in /essd/CK/s3/programs/clickhouse
22. DB::IBlockInputStream::read() @ 0x9c9ec3e in /essd/CK/s3/programs/clickhouse
23. DB::MergeTreeDataMergerMutator::mergePartsToTemporaryPart(DB::FutureMergedMutatedPart const&, std::__1::shared_ptr<DB::StorageInMemoryMetadata const> const&, DB::BackgroundProcessListEntry<DB::MergeListElement, DB::MergeInfo>&, std::__1::shared_ptr<DB::RWLockImpl::LockHolderImpl>&, long, DB::Context const&, std::__1::unique_ptr<DB::IReservation, std::__1::default_delete<DB::IReservation> > const&, bool, std::__1::vector<std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> >, std::__1::allocator<std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > > > const&) @ 0xa3d3d65 in /essd/CK/s3/programs/clickhouse
24. DB::StorageMergeTree::mergeSelectedParts(std::__1::shared_ptr<DB::StorageInMemoryMetadata const> const&, bool, std::__1::vector<std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> >, std::__1::allocator<std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > > > const&, DB::StorageMergeTree::MergeMutateSelectedEntry&, std::__1::shared_ptr<DB::RWLockImpl::LockHolderImpl>&) @ 0xa22dfa7 in /essd/CK/s3/programs/clickhouse
25. bool std::__1::__function::__policy_invoker<bool ()>::__call_impl<std::__1::__function::__default_alloc_func<DB::StorageMergeTree::getDataProcessingJob()::'lambda'(), bool ()> >(std::__1::__function::__policy_storage const*) @ 0xa22e64a in /essd/CK/s3/programs/clickhouse
26. void std::__1::__function::__policy_invoker<void ()>::__call_impl<std::__1::__function::__default_alloc_func<DB::IBackgroundJobExecutor::jobExecutingTask()::'lambda'(), void ()> >(std::__1::__function::__policy_storage const*) @ 0xa33d1b0 in /essd/CK/s3/programs/clickhouse
27. ThreadPoolImpl<ThreadFromGlobalPool>::worker(std::__1::__list_iterator<ThreadFromGlobalPool, void*>) @ 0x3f2f4f3 in /essd/CK/s3/programs/clickhouse
28. ThreadFromGlobalPool::ThreadFromGlobalPool<void ThreadPoolImpl<ThreadFromGlobalPool>::scheduleImpl<void>(std::__1::function<void ()>, int, std::__1::optional<unsigned long>)::'lambda1'()>(void&&, void ThreadPoolImpl<ThreadFromGlobalPool>::scheduleImpl<void>(std::__1::function<void ()>, int, std::__1::optional<unsigned long>)::'lambda1'()&&...)::'lambda'()::operator()() @ 0x3f2fbdf in /essd/CK/s3/programs/clickhouse
29. ThreadPoolImpl<std::__1::thread>::worker(std::__1::__list_iterator<std::__1::thread, void*>) @ 0x3f2eb83 in /essd/CK/s3/programs/clickhouse
30. void* std::__1::__thread_proxy<std::__1::tuple<std::__1::unique_ptr<std::__1::__thread_struct, std::__1::default_delete<std::__1::__thread_struct> >, void ThreadPoolImpl<std::__1::thread>::scheduleImpl<void>(std::__1::function<void ()>, int, std::__1::optional<unsigned long>)::'lambda1'()> >(void*) @ 0x3f2e44f in /essd/CK/s3/programs/clickhouse
31. start_thread @ 0x7ebe in /home/fenglv/gentoo/lib64/libpthread-2.32.so
 (version 21.4.1.1)
```

### PR DESCRIPTION
**Pull request by excitoon**
I hereby agree to the terms of the CLA available at: https://yandex.ru/legal/cla/?lang=en

Changelog category (leave one):
- Improvement

Changelog entry (a user-readable short description of the changes that goes to CHANGELOG.md):

Retries on HTTP connection drops in S3

Fixes #21957

### CODE REVIEW

File: src/Core/Settings.h [hunk 1]
```diff
@@ -70,6 +70,7 @@ class IColumn;
     M(UInt64, connections_with_failover_max_tries, DBMS_CONNECTION_POOL_WITH_FAILOVER_DEFAULT_MAX_TRIES, "The maximum number of attempts to connect to replicas.", 0) \
     M(UInt64, s3_min_upload_part_size, 512*1024*1024, "The minimum size of part to upload during multipart upload to S3.", 0) \
     M(UInt64, s3_max_single_part_upload_size, 64*1024*1024, "The maximum size of object to upload using singlepart upload to S3.", 0) \
+    M(Int64, s3_max_single_read_retries, 16, "The maximum number of retries during single S3 read.", 0) \
```
Comment (nikitamikhaylov) [inline src/Core/Settings.h:73]: Why Int64, not UInt64 ?

Comment (nikitamikhaylov) [reply src/Core/Settings.h:73]: And why it is set to 16 by default? I think it is too many.

**Review by nikitamikhaylov — APPROVED** [review]

File: src/Core/Settings.h (hunk 1 above)
Comment (excitoon) [reply src/Core/Settings.h:73]: I wanted to have possibility to have infinite retries, by specifying -1. Will 4 be okay? @nikitamikhaylov

Comment (nikitamikhaylov) [reply src/Core/Settings.h:73]: @excitoon Infinite retries is the worth thing ever. 4 retries is Ok.

Comment (excitoon) [reply src/Core/Settings.h:73]: +

**Review by nikitamikhaylov — APPROVED** [review]

**Comment by alexey-milovidov** [comment]
We decided that exponential backoff is needed.

**Comment by excitoon** [comment]
@alexey-milovidov okay I will make it in separate PR.

**Comment by excitoon** [comment]
@alexey-milovidov @nikitamikhaylov Please look at #23461

**Comment by ucasFL** [comment]
This doesn't fixed my problem #21957. 

I'm a little confused about the `ReadBuffer` implementation.  

I use another naive way to solve this problem, which works well: I use one variable named `already_read` both in `ReadBufferFromIStream` and `ReadBufferFromS3` to record already read bytes, when exception happened, plus `aready_read` value in `ReadBufferFromIStream` to `already_read` in `ReadBufferFromS3`, and use `new_offset = offset + already_read` to reinitialize `ReadBufferFromIStream`.  The code is in here: https://github.com/ucasFL/s3/tree/master/master, and LOG shows that `new_offset` is not equal to `getPosition()`:

```
2021.05.03 23:13:07.365468 [ 17469 ] {} <Debug> ReadBufferFromS3: DIFF: getPosition: 0, new_offset:1048576
2021.05.03 23:13:12.601939 [ 17469 ] {} <Debug> ReadBufferFromS3: DIFF: getPosition: 0, new_offset:2097152
2021.05.03 23:13:12.902443 [ 17337 ] {} <Debug> ReadBufferFromS3: DIFF: getPosition: 0, new_offset:1048576
2021.05.03 23:13:13.127227 [ 17362 ] {} <Debug> ReadBufferFromS3: DIFF: getPosition: 0, new_offset:1048576
2021.05.03 23:13:21.791544 [ 17377 ] {} <Debug> ReadBufferFromS3: DIFF: getPosition: 0, new_offset:2097152
2021.05.03 23:13:39.747449 [ 17386 ] {} <Debug> ReadBufferFromS3: DIFF: getPosition: 0, new_offset:1048576
2021.05.03 23:13:48.079606 [ 17377 ] {} <Debug> ReadBufferFromS3: DIFF: getPosition: 0, new_offset:1048576
2021.05.03 23:13:51.995675 [ 17302 ] {e4bc213c-d42d-4a60-b844-96888b50ffb2} <Debug> ReadBufferFromS3: DIFF: getPosition: 0, new_offset:50331648
2021.05.03 23:13:55.706082 [ 17309 ] {e4bc213c-d42d-4a60-b844-96888b50ffb2} <Debug> ReadBufferFromS3: DIFF: getPosition: 305256658, new_offset:475125970
2021.05.03 23:14:26.053909 [ 17334 ] {e4bc213c-d42d-4a60-b844-96888b50ffb2} <Debug> ReadBufferFromS3: DIFF: getPosition: 0, new_offset:1048576
2021.05.03 23:14:53.194259 [ 17369 ] {e4bc213c-d42d-4a60-b844-96888b50ffb2} <Debug> ReadBufferFromS3: DIFF: getPosition: 1238873065, new_offset:1948759017
2021.05.03 23:15:16.047457 [ 17312 ] {e4bc213c-d42d-4a60-b844-96888b50ffb2} <Debug> ReadBufferFromS3: DIFF: getPosition: 0, new_offset:1048576
2021.05.03 23:15:44.693221 [ 17407 ] {e4bc213c-d42d-4a60-b844-96888b50ffb2} <Debug> ReadBufferFromS3: DIFF: getPosition: 0, new_offset:1048576
2021.05.03 23:16:12.878783 [ 17451 ] {} <Debug> ReadBufferFromS3: DIFF: getPosition: 0, new_offset:5242880
2021.05.03 23:16:14.047713 [ 17331 ] {e4bc213c-d42d-4a60-b844-96888b50ffb2} <Debug> ReadBufferFromS3: DIFF: getPosition: 0, new_offset:257949696
2021.05.03 23:16:47.841742 [ 17413 ] {e4bc213c-d42d-4a60-b844-96888b50ffb2} <Debug> ReadBufferFromS3: DIFF: getPosition: 0, new_offset:1048576
2021.05.03 23:16:53.695540 [ 17327 ] {e4bc213c-d42d-4a60-b844-96888b50ffb2} <Debug> ReadBufferFromS3: DIFF: getPosition: 32841844, new_offset:108339316
2021.05.03 23:16:59.928206 [ 17329 ] {e4bc213c-d42d-4a60-b844-96888b50ffb2} <Debug> ReadBufferFromS3: DIFF: getPosition: 0, new_offset:1048576
2021.05.03 23:17:15.945801 [ 17468 ] {e4bc213c-d42d-4a60-b844-96888b50ffb2} <Debug> ReadBufferFromS3: DIFF: getPosition: 0, new_offset:1048576
2021.05.03 23:17:33.434971 [ 17403 ] {e4bc213c-d42d-4a60-b844-96888b50ffb2} <Debug> ReadBufferFromS3: DIFF: getPosition: 0, new_offset:1048576
2021.05.03 23:17:38.643761 [ 17426 ] {e4bc213c-d42d-4a60-b844-96888b50ffb2} <Debug> ReadBufferFromS3: DIFF: getPosition: 0, new_offset:1048576
2021.05.03 23:20:03.287400 [ 17451 ] {} <Debug> ReadBufferFromS3: DIFF: getPosition: 0, new_offset:12582912
2021.05.03 23:20:57.414397 [ 17362 ] {} <Debug> ReadBufferFromS3: DIFF: getPosition: 0, new_offset:380633088
2021.05.03 23:23:03.617168 [ 17451 ] {} <Debug> ReadBufferFromS3: DIFF: getPosition: 0, new_offset:12582912
2021.05.03 23:24:51.062112 [ 17362 ] {} <Debug> ReadBufferFromS3: DIFF: getPosition: 0, new_offset:233832448
2021.05.03 23:25:00.327010 [ 17362 ] {} <Debug> ReadBufferFromS3: DIFF: getPosition: 0, new_offset:855638016

```

File: src/IO/ReadBufferFromS3.cpp [hunk 2]
```diff
@@ -29,26 +31,58 @@ namespace ErrorCodes
...
+        try
+        {
+            next_result = impl->next();
+            /// FIXME. 1. Poco `istream` cannot read less than buffer_size or this state is being discarded during
+            ///           istream <-> iostream conversion. `gcount` always contains 0,
+            ///           that's why we always have error "Cannot read from istream at offset 0".
+
+            break;
+        }
+        catch (const Exception & e)
```
Comment (azat) [inline src/IO/ReadBufferFromS3.cpp:69]: Catching **any** exception does not looks sane, @excitoon what do you think about enumerate codes that should be retriable?

**Pull request merged**